pkg/ccl/importccl/read_import_pgdump.go

```diff
@@ -102,3 +108,212 @@ func (p *postgreStream) read() (tree.StatementList, error) {
 		}
 	}
 }
+
+func readPostgresCreateTable(
+	input io.Reader, evalCtx *tree.EvalContext, settings *cluster.Settings, match string, parentID sqlbase.ID, walltime int64,
+) ([]*sqlbase.TableDescriptor, error) {
+
+	// Modify the CreateTable stmt with the various index additions. We do this
+	// instead of creating a full table descriptor first and adding indexes
+	// later because MakeSimpleTableDescriptor calls the sql package which calls
+	// AllocateIDs which adds the hidden rowid and default primary key. This means
+	// we'd have to delete the index and row and modify the column family. This
+	// is much easier and probably safer too.
+	createTbl := make(map[string]*tree.CreateTable)
+	var found []string
+	ps := newPostgreStream(input, defaultPgStreamInit, defaultPgStreamMax)
+	for {
+		stmt, err := ps.Next()
+		if err == io.EOF {
+			if len(createTbl) == 0 {
+				return nil, errors.Errorf("no table definition found")
+			}
+			ret := make([]*sqlbase.TableDescriptor, 0, len(createTbl))
+			for name, create := range createTbl {
+				if match == "" || match == name {
+					id := sqlbase.ID(int(defaultCSVTableID) + len(ret))
+					desc, err := MakeSimpleTableDescriptor(evalCtx.Ctx(), settings, create, parentID, id, walltime)
+					if err != nil {
+						return nil, err
+					}
+					ret = append(ret, desc)
+				}
+			}
+			if match != "" && len(ret) != 1 {
+				return nil, errors.Errorf("table %q not found in file (found tables: %s)", match, strings.Join(found, ", "))
+			}
+			return ret, nil
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "postgres parse error")
+		}
+		switch stmt := stmt.(type) {
+		case *tree.CreateTable:
+			name, err := getTableName(stmt.Table)
+			if err != nil {
+				return nil, err
+			}
+			if match != "" && match != name {
+				found = append(found, name)
+				continue
+			}
+			createTbl[name] = stmt
+		case *tree.CreateIndex:
+			name, err := getTableName(stmt.Table)
+			if err != nil {
+				return nil, err
+			}
+			create, ok := createTbl[name]
+			if !ok {
+				continue
+			}
+			var idx tree.TableDef = &tree.IndexTableDef{
+				Name:        stmt.Name,
+				Columns:     stmt.Columns,
+				Storing:     stmt.Storing,
+				Inverted:    stmt.Inverted,
+				Interleave:  stmt.Interleave,
+				PartitionBy: stmt.PartitionBy,
+			}
+			if stmt.Unique {
+				idx = &tree.UniqueConstraintTableDef{IndexTableDef: *idx.(*tree.IndexTableDef)}
+			}
+			create.Defs = append(create.Defs, idx)
+		case *tree.AlterTable:
+			name, err := getTableName(stmt.Table)
+			if err != nil {
+				return nil, err
+			}
+			create, ok := createTbl[name]
+			if !ok {
+				continue
```

Comment:
hm. if you go the nil/non-nil map entry route, I'd error on `!ok`, and only `continue` on nil.

Reply:
i've changed this to continue on nil, no error path. i kinda don't care if there's a stray alter table statement for which we haven't seen an associated create.